addons/addon-base-raas/packages/base-raas-services/lib/data-egress/data-egress-service.js

```diff
@@ -124,15 +142,124 @@ class DataEgressService extends Service {
         },
       ],
     };
+    const memberAccountId = await this.getMemberAccountId(requestContext, environment.id);
 
-    const environmentScService = await this.service('environmentScService');
-    const environmentScEntity = await environmentScService.mustFind(requestContext, { id: environment.id });
-    const memberAccount = await environmentScService.getMemberAccount(requestContext, environmentScEntity);
-    await this.addEgressStoreBucketPolicy(requestContext, egressStore, memberAccount.accountId);
+    const bucketPolicyLockId = `bucket-policy-access-${bucketName}`;
+    await lockService.tryWriteLockAndRun({ id: bucketPolicyLockId }, async () => {
+      await this.addEgressStoreBucketPolicy(requestContext, egressStore, memberAccountId);
+    });
 
     return egressStore;
   }
 
+  async terminateEgressStore(requestContext, environmentId) {
+    const enableEgressStore = this.settings.get(settingKeys.enableEgressStore);
+    const curUser = _.get(requestContext, 'principalIdentifier.uid');
+    if (!enableEgressStore) {
+      throw this.boom.forbidden('Unable to create Egress store since this feature is disabled', true);
+    }
+
+    const workspaceId = environmentId;
+    let egressStoreScanResult = [];
+
+    try {
+      egressStoreScanResult = await this._scanner()
+        .limit(1000)
+        .scan()
+        .then(egressStores => {
+          return egressStores.filter(store => store.workspaceId === workspaceId);
+        });
+    } catch (error) {
+      throw this.boom.notFound(`Error in terminating egress store: ${JSON.stringify(error)}`, true);
+    }
+
+    if (egressStoreScanResult.length !== 1) {
+      throw this.boom.internalError(
+        `Error in terminating egress store: multiple result fetched from egrss store table`,
+        true,
+      );
+    }
+    const egressStoreInfo = egressStoreScanResult[0];
+    const isEgressStoreOwner = egressStoreInfo.createdBy === curUser;
+    if (!isAdmin(requestContext) && !isEgressStoreOwner) {
+      throw this.boom.forbidden(
+        `You are not authorized to terminate the egress store. Please contact your administrator.`,
+        true,
+      );
+    }
+
+    const s3Service = await this.service('s3Service');
+    const egressStoreStatus = egressStoreInfo.status;
+
+    if (egressStoreStatus.toUpperCase() === PROCESSING_STATUS_CODE) {
```

Comment:
If the store is in CREATED state, I think we should throw an exception stating that store doesn't have the right state to be terminated since it hasn't processed yet